test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/RazorGenerateIntegrationTest.cs

```diff
@@ -52,5 +57,126 @@ public class RazorGenerateIntegrationTest : MSBuildIntegrationTestBase
             // The file should still be generated even if we had a Razor syntax error.
             Assert.FileExists(result, RazorIntermediateOutputPath, "Views", "Home", "Index.cs");
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorGenerate_BuildsIncrementally()
+        {
+            // Act - 1
+            var result = await DotnetMSBuild(RazorGenerateTarget);
+            var generatedFile = Path.Combine(Project.DirectoryPath, RazorIntermediateOutputPath, "Views", "Home", "About.cs");
+
+            // Assert - 1
+            Assert.BuildPassed(result);
+            Assert.FileExists(result, generatedFile);
+            var thumbPrint = GetThumbPrint(generatedFile);
+
+            // Act - 2
+            using (var razorGenDirectoryLock = LockDirectory(RazorIntermediateOutputPath))
+            {
+                result = await DotnetMSBuild(RazorGenerateTarget);
+            }
+
+            // Assert - 2
+            Assert.BuildPassed(result);
+            Assert.FileExists(result, generatedFile);
+            var currentThumbPrint = GetThumbPrint(generatedFile);
+            Assert.Equal(thumbPrint, currentThumbPrint);
+        }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorGenerate_Rebuilds_IfSourcesAreUpdated()
+        {
+            // Act - 1
+            var result = await DotnetMSBuild(RazorGenerateTarget);
+            var file = Path.Combine(Project.DirectoryPath, "Views", "Home", "Contact.cshtml");
+            var generatedFile = Path.Combine(RazorIntermediateOutputPath, "Views", "Home", "Contact.cs");
+
+            // Assert - 1
+            Assert.BuildPassed(result);
+            var fileThumbPrint = GetThumbPrint(generatedFile);
+
+            // Act - 2
+            // Update the source content and build. We should expect the outputs to be regenerated.
+            // Timestamps on xplat are precise only to a second. Add a delay so we can ensure that MSBuild recognizes the
+            // file change. See https://github.com/dotnet/corefx/issues/26024
+            await Task.Delay(TimeSpan.FromSeconds(1));
```

Comment:
I don't feel good about this. We should use something like this as a workaround instead https://msdn.microsoft.com/en-us/library/system.io.file.setlastwritetime(v=vs.110).aspx

Reply:
They do have a 1 second granularity, so this works pretty reliably. Setting a write time in the future, which is sort of what we'd do have to effectively do to get the roll up is a poorer approximation of what a user would do.

Reply:
Delays in tests are fragile, and don't scale well.